wallet-next/src/bin/pwalletd.rs

```diff
@@ -1,22 +1,210 @@
+            //         })
+            //         .add_service(WalletServer::new(storage.clone()))
+            //         .serve(
+            //             format!("{}:{}", host, wallet_query_port)
+            //                 .parse()
+            //                 .expect("this is a valid address"),
+            //         ),
+            // );
+        }
+    }
+
+    Ok(())
+}
+
+pub async fn sync(client: &ObliviousQueryClient<Channel>, storage: &mut Storage) -> Result<()> {
+    tracing::info!("starting client sync");
+
+    let start_height = storage
+        .last_block_height()
+        .await
+        .map(|h| h + 1)
+        .unwrap_or(0);
+    let mut stream = client
+        .compact_block_range(tonic::Request::new(CompactBlockRangeRequest {
+            start_height,
+            end_height: 0,
+            chain_id: storage
+                .chain_id()
+                .await
+                .ok_or_else(|| anyhow::anyhow!("missing chain_id"))?,
+        }))
+        .await?
+        .into_inner();
+
+    let mut count = 0;
+    while let Some(block) = stream.message().await? {
+        storage.scan_block(block.try_into()?)?;
+        // very basic form of intermediate checkpointing
+        count += 1;
+        if count % 1000 == 1 {
+            storage.commit()?;
+            tracing::info!(height = ?storage.last_block_height().await.unwrap(), "syncing...");
+        }
+    }
+
+    storage.prune_timeouts();
+    storage.commit()?;
+    tracing::info!(end_height = ?storage.last_block_height().await.unwrap(), "finished sync");
+    Ok(())
+}
+
+pub async fn assets(client: &ObliviousQueryClient<Channel>, storage: &mut Storage) -> Result<()> {
+    // Update asset registry.
+    let request = tonic::Request::new(AssetListRequest {
+        chain_id: storage.chain_id().await.unwrap_or_default(),
+    });
+    let assets: KnownAssets = client.asset_list(request).await?.into_inner().try_into()?;
+    for asset in assets.0 {
+        storage
+            .asset_cache_mut()
+            .extend(std::iter::once(asset.denom));
+    }
+
+    storage.commit()?;
+    tracing::info!("updated asset registry");
+    Ok(())
+}
+
+/// Fetches the global chain parameters and stores them on `storage`.
 
-    storage.migrate().await?;
+pub async fn chain_params(
+    client: &ObliviousQueryClient<Channel>,
+    storage: &mut Storage,
+) -> Result<()> {
+    let params = client
+        .chain_params(tonic::Request::new(ChainParamsRequest {
+            chain_id: storage.chain_id().await.unwrap_or_default(),
+        }))
+        .await?
+        .into_inner()
+        .into();
 
-    let row = storage.insert_table().await?;
-    let x = storage.read_table().await?;
+    tracing::info!(?params, "saving chain params");
 
-    println!(
-        "Hello, pwalletd! I got stuff from sqlite: row {} value {}",
-        row, x
-    );
+    *storage.chain_params_mut() = Some(params);
+    storage.commit()?;
     Ok(())
 }
```

Comment:
This method is also probably deletable; we'll need to get the chain params, but we won't want to do so here, and the API this implementation uses (using a `_mut()` method to get a mutable reference, setting a value there, and then calling a commit function) isn't compatible with the state model we're using (SQLite queries).  So there's not much code to save for reuse later other than the `client.chain_params` invocation, which we can redo whenever.